File: daemon-dart-archive/lib/api/lora_api.dart
import 'dart:convert';
import 'package:shelf/shelf.dart' as shelf;
import 'package:shelf_router/shelf_router.dart';
import '../database/app_database.dart';

/// REST API for LoRA registry and generation recipe management.
class LoRAApi {
  final AppDatabase _db;

  LoRAApi(this._db);

  void registerRoutes(Router router) {
    // LoRA registry
    router.get('/api/v1/loras', _handleListLoRAs);
    router.get('/api/v1/loras/<id>', _handleGetLoRA);
    router.post('/api/v1/loras', _handleCreateLoRA);
    router.put('/api/v1/loras/<id>', _handleUpdateLoRA);
    router.delete('/api/v1/loras/<id>', _handleDeleteLoRA);

    // Generation recipes
    router.get('/api/v1/recipes', _handleListRecipes);
    router.get('/api/v1/recipes/<id>', _handleGetRecipe);
    router.post('/api/v1/recipes', _handleCreateRecipe);
    router.put('/api/v1/recipes/<id>', _handleUpdateRecipe);
    router.delete('/api/v1/recipes/<id>', _handleDeleteRecipe);
  }

  // ── LoRA endpoints ──

  Future<shelf.Response> _handleListLoRAs(shelf.Request request) async {
    final type = request.url.queryParameters['type'];
    final loras = await _db.listLoRAs(type: type);
    return _json({'success': true, 'loras': loras});
  }

  Future<shelf.Response> _handleGetLoRA(shelf.Request request, String id) async {
    final lora = await _db.getLoRA(id);
    if (lora == null) {
      return _json({'success': false, 'error': 'LoRA not found'}, 404);
    }
    return _json({'success': true, 'lora': lora});
  }

  Future<shelf.Response> _handleCreateLoRA(shelf.Request request) async {
    try {
      final body = await request.readAsString();
      final json = jsonDecode(body) as Map<String, dynamic>;

      final now = DateTime.now().millisecondsSinceEpoch;
      final id = json['id'] as String? ?? 'lora_$now';

      await _db.upsertLoRA({
        'id': id,
        'name': json['name'] ?? id,
        'type': json['type'] ?? 'style',
        'path': json['path'] ?? '',
        'trigger_word': json['trigger_word'] ?? '',
        'weight': (json['weight'] as num?)?.toDouble() ?? 0.7,
        'preview_base64': json['preview_base64'],
        'tags': jsonEncode(json['tags'] ?? []),
        'created_at': now,
      });

      return _json({'success': true, 'id': id}, 201);
    } catch (e) {
      return _json({'success': false, 'error': 'Failed to create LoRA: $e'}, 400);
    }
  }

  Future<shelf.Response> _handleUpdateLoRA(shelf.Request request, String id) async {
    try {
      final existing = await _db.getLoRA(id);
      if (existing == null) {
        return _json({'success': false, 'error': 'LoRA not found'}, 404);
      }

      final body = await request.readAsString();
      final json = jsonDecode(body) as Map<String, dynamic>;

      final updated = Map<String, dynamic>.from(existing);
      if (json.containsKey('name')) updated['name'] = json['name'];
      if (json.containsKey('type')) updated['type'] = json['type'];
      if (json.containsKey('path')) updated['path'] = json['path'];
      if (json.containsKey('trigger_word')) updated['trigger_word'] = json['trigger_word'];
      if (json.containsKey('weight')) updated['weight'] = (json['weight'] as num).toDouble();
      if (json.containsKey('tags')) updated['tags'] = jsonEncode(json['tags']);
      if (json.containsKey('preview_base64')) updated['preview_base64'] = json['preview_base64'];

      await _db.upsertLoRA(updated);
      return _json({'success': true, 'lora': updated});
    } catch (e) {
      return _json({'success': false, 'error': 'Update failed: $e'}, 400);
    }
  }

  Future<shelf.Response> _handleDeleteLoRA(shelf.Request request, String id) async {
    final deleted = await _db.deleteLoRA(id);
    if (!deleted) {
      return _json({'success': false, 'error': 'LoRA not found'}, 404);
    }
    return _json({'success': true});
  }

  // ── Recipe endpoints ──

  Future<shelf.Response> _handleListRecipes(shelf.Request request) async {
    final recipes = await _db.listRecipes();
    return _json({'success': true, 'recipes': recipes});
  }

  Future<shelf.Response> _handleGetRecipe(shelf.Request request, String id) async {
    final recipe = await _db.getRecipe(id);
    if (recipe == null) {
      return _json({'success': false, 'error': 'Recipe not found'}, 404);
    }
    return _json({'success': true, 'recipe': recipe});
  }

  Future<shelf.Response> _handleCreateRecipe(shelf.Request request) async {
    try {
      final body = await request.readAsString();
      final json = jsonDecode(body) as Map<String, dynamic>;

      final now = DateTime.now().millisecondsSinceEpoch;
      final id = json['id'] as String? ?? 'recipe_$now';

      await _db.upsertRecipe({
        'id': id,
        'name': json['name'] ?? 'Untitled Recipe',
        'description': json['description'] ?? '',
        'image_model': json['image_model'] ?? 'animagine_xl',
        'video_model': json['video_model'] ?? 'local_v3',
        'quality': json['quality'] ?? 'standard',
        'lora_ids': jsonEncode(json['lora_ids'] ?? []),
        'controlnet_type': json['controlnet_type'] ?? 'lineart_anime',
        'controlnet_scale': (json['controlnet_scale'] as num?)?.toDouble() ?? 0.7,
        'ip_adapter_scale': (json['ip_adapter_scale'] as num?)?.toDouble() ?? 0.6,
        'color_grade': json['color_grade'] ?? '',
        'export_platform': json['export_platform'] ?? '',
        'created_at': now,
        'updated_at': now,
      });

      return _json({'success': true, 'id': id}, 201);
    } catch (e) {
      return _json({'success': false, 'error': 'Failed to create recipe: $e'}, 400);
    }
  }

  Future<shelf.Response> _handleUpdateRecipe(shelf.Request request, String id) async {
    try {
      final existing = await _db.getRecipe(id);
      if (existing == null) {
        return _json({'success': false, 'error': 'Recipe not found'}, 404);
      }

      final body = await request.readAsString();
      final json = jsonDecode(body) as Map<String, dynamic>;

      final updated = Map<String, dynamic>.from(existing);
      for (final key in ['name', 'description', 'image_model', 'video_model',
          'quality', 'controlnet_type', 'color_grade', 'export_platform']) {
        if (json.containsKey(key)) updated[key] = json[key];
      }
      if (json.containsKey('lora_ids')) updated['lora_ids'] = jsonEncode(json['lora_ids']);
      if (json.containsKey('controlnet_scale')) {
        updated['controlnet_scale'] = (json['controlnet_scale'] as num).toDouble();
      }
      if (json.containsKey('ip_adapter_scale')) {
        updated['ip_adapter_scale'] = (json['ip_adapter_scale'] as num).toDouble();
      }
      updated['updated_at'] = DateTime.now().millisecondsSinceEpoch;

      await _db.upsertRecipe(updated);
      return _json({'success': true, 'recipe': updated});
    } catch (e) {
      return _json({'success': false, 'error': 'Update failed: $e'}, 400);
    }
  }

  Future<shelf.Response> _handleDeleteRecipe(shelf.Request request, String id) async {
    final deleted = await _db.deleteRecipe(id);
    if (!deleted) {
      return _json({'success': false, 'error': 'Recipe not found'}, 404);
    }
    return _json({'success': true});
  }

  shelf.Response _json(Map<String, dynamic> data, [int status = 200]) {
    return shelf.Response(status,
        body: jsonEncode(data),
        headers: {'Content-Type': 'application/json'});
  }
}

File: daemon/opencli_daemon/api/lora_api.py
"""LoRA + Recipe CRUD API.

Ported from daemon/lib/api/lora_api.dart.
"""

import time
import uuid

from fastapi import APIRouter, Request

from opencli_daemon.database import connection as db

router = APIRouter(prefix="/api/v1", tags=["lora"])


# ── LoRA Registry ────────────────────────────────────────────────────────────


@router.get("/loras")
async def list_loras(type: str | None = None) -> dict:
    if type:
        rows = await db.list_rows("lora_registry", where="type = ?", params=(type,),
                                  order_by="created_at DESC")
    else:
        rows = await db.list_rows("lora_registry", order_by="created_at DESC")
    return {"loras": rows}


@router.get("/loras/{lora_id}")
async def get_lora(lora_id: str) -> dict:
    row = await db.get_row("lora_registry", "id", lora_id)
    if row is None:
        return {"error": f"LoRA not found: {lora_id}"}
    return {"lora": row}


@router.post("/loras")
async def create_lora(request: Request) -> dict:
    body = await request.json()
    lid = body.get("id", str(uuid.uuid4()))
    now = int(time.time() * 1000)
    await db.upsert_row("lora_registry", {
        "id": lid,
        "name": body.get("name", ""),
        "type": body.get("type", "style"),
        "path": body.get("path", ""),
        "trigger_word": body.get("trigger_word", ""),
        "weight": body.get("weight", 0.7),
        "preview_base64": body.get("preview_base64"),
        "tags": body.get("tags", "[]") if isinstance(body.get("tags"), str) else str(body.get("tags", "[]")),
        "created_at": now,
    })
    return {"success": True, "id": lid}


@router.delete("/loras/{lora_id}")
async def delete_lora(lora_id: str) -> dict:
    deleted = await db.delete_row("lora_registry", "id", lora_id)
    return {"success": deleted}


# ── Generation Recipes ───────────────────────────────────────────────────────


@router.get("/recipes")
async def list_recipes() -> dict:
    rows = await db.list_rows("generation_recipes", order_by="updated_at DESC")
    return {"recipes": rows}


@router.get("/recipes/{recipe_id}")
async def get_recipe(recipe_id: str) -> dict:
    row = await db.get_row("generation_recipes", "id", recipe_id)
    if row is None:
        return {"error": f"Recipe not found: {recipe_id}"}
    return {"recipe": row}


@router.post("/recipes")
async def create_recipe(request: Request) -> dict:
    body = await request.json()
    rid = body.get("id", str(uuid.uuid4()))
    now = int(time.time() * 1000)
    await db.upsert_row("generation_recipes", {
        "id": rid,
        "name": body.get("name", ""),
        "description": body.get("description", ""),
        "image_model": body.get("image_model", "animagine_xl"),
        "video_model": body.get("video_model", "local_v3"),
        "quality": body.get("quality", "standard"),
        "lora_ids": body.get("lora_ids", "[]") if isinstance(body.get("lora_ids"), str) else str(body.get("lora_ids", "[]")),
        "controlnet_type": body.get("controlnet_type", "lineart_anime"),
        "controlnet_scale": body.get("controlnet_scale", 0.7),
        "ip_adapter_scale": body.get("ip_adapter_scale", 0.6),
        "color_grade": body.get("color_grade", ""),
        "export_platform": body.get("export_platform", ""),
        "created_at": now,
        "updated_at": now,
    })
    return {"success": True, "id": rid}


@router.delete("/recipes/{recipe_id}")
async def delete_recipe(recipe_id: str) -> dict:
    deleted = await db.delete_row("generation_recipes", "id", recipe_id)
    return {"success": deleted}
